core/desugaring.cpp

```diff
@@ -511,13 +511,47 @@ class Desugarer {
             // Nothing to do.
 
         } else if (auto *ast = dynamic_cast<Index*>(ast_)) {
-            desugar(ast->target, obj_level);
-            if (ast->id != nullptr) {
-                assert(ast->index == nullptr);
-                ast->index = str(ast->id->name);
-                ast->id = nullptr;
+            if (ast->isSlice) {
+                if (ast->end != nullptr) {
+                    desugar(ast->end, obj_level);
+                } else {
+                    ast->end = length(ast->target);
```

Comment:
If you desugared them to null, you could test for null in std.slice and implement the full semantics that way